cf-custom-resources/lib/custom-domain.js

```diff
@@ -0,0 +1,360 @@
+ * @param {string} physicalResourceId CloudFormation physical resource ID
+ * @param {object} [responseData] arbitrary response data object
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+let report = function (
+  event,
+  context,
+  responseStatus,
+  physicalResourceId,
+  responseData,
+  reason
+) {
+  return new Promise((resolve, reject) => {
+    const https = require("https");
+    const { URL } = require("url");
+
+    var responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: reason,
+      PhysicalResourceId: physicalResourceId || context.logStreamName,
+      StackId: event.StackId,
+      RequestId: event.RequestId,
+      LogicalResourceId: event.LogicalResourceId,
+      Data: responseData,
+    });
+
+    const parsedUrl = new URL(event.ResponseURL || defaultResponseURL);
+    const options = {
+      hostname: parsedUrl.hostname,
+      port: 443,
+      path: parsedUrl.pathname + parsedUrl.search,
+      method: "PUT",
+      headers: {
+        "Content-Type": "",
+        "Content-Length": responseBody.length,
+      },
+    };
+
+    https
+      .request(options)
+      .on("error", reject)
+      .on("response", (res) => {
+        res.resume();
+        if (res.statusCode >= 400) {
+          reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+        } else {
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+};
+
+/**
+ * Upsert all alias records to the correct domain hosted zone. More specifically,
+ * we'll add the record to the root hosted zone for aliases in format of `*.${domainName}`;
+ * to the app hosted zone for aliases in format of `*.$appName}.${domainName}`;
+ * and to the env hosted zone for aliases in format of `*.${envName}.${appName}.${domainName}`.
+ * Also for the other aliases not matching any of the condition above, we'll skip since
+ * it is corresponding hosted zone is not managable by Copilot.
+ *
+ * @param {string} aliases the custom domain aliases
+ * @param {string} lbDNS DNS of the load balancer
+ * @param {string} lbHostedZone Hosted Zone of the load balancer
+ * @param {string} rootDnsRole the IAM role ARN that can manage domainName
+ * @param {string} aliasTypes the alias type
+ */
+const writeCustomDomainRecord = async function (
+  aliases,
+  lbDNS,
+  lbHostedZone,
+  rootDnsRole,
+  aliasTypes,
+  action
+) {
+  const envRoute53 = new aws.Route53();
+  const appRoute53 = new aws.Route53({
+    credentials: new aws.ChainableTemporaryCredentials({
+      params: { RoleArn: rootDnsRole },
+      masterCredentials: new aws.EnvironmentCredentials("AWS"),
+    }),
+  });
+  if (waiter) {
+    // Used by the test suite, since waiters aren't mockable yet
+    envRoute53.waitFor = appRoute53.waitFor = waiter;
+  }
```

Comment:
can we define these as singletons so that we don't recreate clients for each alias we need to update?

Would it make it difficult to test if we declare these clients outside of the function handler:
```js
let envRoute53, appRoute53;

exports.handler = .. => {
   envRoute53 = new aws.Route53();
   ...
}
```

Reply:
Since we won't be creating those clients for each alias, it will be only created once for create and delete and twice for update which I think should be fine? Updated tho so that we only create those clients once even for update.